modules/nextflow/src/main/groovy/nextflow/file/FilePorter.groovy

```diff
@@ -194,8 +212,11 @@ class FilePorter {
          */
         Path addToForeign(Path path) {
             // copy the path with a thread pool
-            foreignPaths << path
-            return getCachePathFor(path, stageDir)
+            synchronized (owner) {
+                final target = getCachePathFor(path, stageDir)
```

Comment:
This PR will fix the problem but it will download the same files twice, because still it's possible to call this `getCachePathFor` while a transfer is running on the background.

Maybe here we can check if there is already a `FileCopy` at the `FilePorter.stagingTransfers` for the current `path` and reuse it instead of calling a new `getCachePathFor` that it may return a different one.

Reply:
Not sure, as long as the `FileCopy` identity is made by both source and target paths, if a second copy is created for the same path it will receive the same transfer object here 

https://github.com/nextflow-io/nextflow/blob/4bfbc7e6426016341cdd30e591d372a08f1f34ab/modules/nextflow/src/main/groovy/nextflow/file/FilePorter.groovy#L99-L110

Reply:
and this could happen if there's a race condition when checking the existence of the target path and both return false

Reply:
but it's true that if the check is made while the transfer is on-going it will cause of duplicate download

Reply:
the problem is that it's hard to know if it's an ongoing download or a broken/partial copy left by another run.

Reply:
But can another run be at `FilePorter.stagingTransfers` ? or it has only the ones from the current one?

Reply:
Do you mean another nextflow run? it may happen if the work dir is the same 

https://github.com/nextflow-io/nextflow/blob/4bfbc7e6426016341cdd30e591d372a08f1f34ab/modules/nextflow/src/main/groovy/nextflow/executor/Executor.groovy#L120-L120

This can be improved by adding the session id to the stage path. Something like 

```
getWorkDir().resolve("stage-$sessionId")
```

Reply:
Need to fix the tests, but what about this 👉 [b91629d](https://github.com/nextflow-io/nextflow/pull/3511/commits/b91629d7250e5fe34da21247ec6e24f37ddd5b19)